Docker Image v1.78.1: SIGSEGV: segmentation violation

### What is the issue?

A few of my services in containers running on a small server with Docker Compose running Tailscale as a sidecar went down a few hours ago.  I also had [Watchtower](https://github.com/containrrr/watchtower) configured in Docker Compose to update the images. As a last resort after some investigation and troubleshooting I switched the definition from `image: tailscale/tailscale:latest` to `image: tailscale/tailscale:v1.76` which revived the first service I have running in Docker Compose.

As far as I can see the time when the services went down correlates with the image built and published:
https://hub.docker.com/layers/tailscale/tailscale/latest/images/sha256-69203be03a0d30bd23acaccdecfbfb41cd21d9e4997e305ad47b881e4e0acaa3?context=explore

I'm aware that correlation does not imply causation, but please kindly look into this and advise.

```
boot: 2024/12/10 23:40:55 Starting tailscaled
boot: 2024/12/10 23:40:55 Waiting for tailscaled socket
2024/12/10 23:40:55 logtail started
2024/12/10 23:40:55 Program starting: v1.78.1-tc80eb698d, Go 1.23.3: []string{"tailscaled", "--socket=/tmp/tailscaled.sock", "--statedir=/var/lib/tailscale"}
2024/12/10 23:40:55 LogID: b81787e5ddb79b647317d01a7e310fad69865d61b22803ccdba8aea3e19d4cbc
2024/12/10 23:40:55 logpolicy: using system state directory "/var/lib/tailscale"
2024/12/10 23:40:55 dns: [rc=unknown ret=direct]
2024/12/10 23:40:55 dns: using "direct" mode
2024/12/10 23:40:55 dns: using *dns.directManager
2024/12/10 23:40:55 dns: inotify addwatch: context canceled
2024/12/10 23:40:55 wgengine.NewUserspaceEngine(tun "tailscale0") ...
2024/12/10 23:40:55 dns: [rc=unknown ret=direct]
2024/12/10 23:40:55 dns: using "direct" mode
2024/12/10 23:40:55 dns: using *dns.directManager
2024/12/10 23:40:55 link state: interfaces.State{defaultRoute=eth0 ifs={eth0:[172.19.0.2/16]} v4=true v6=false}
2024/12/10 23:40:55 onPortUpdate(port=41366, network=udp6)
2024/12/10 23:40:55 router: using firewall mode pref 
2024/12/10 23:40:55 router: default choosing iptables
2024/12/10 23:40:55 router: netfilter running in iptables mode v6 = true, v6filter = true, v6nat = true
2024/12/10 23:40:55 onPortUpdate(port=42519, network=udp4)
2024/12/10 23:40:55 magicsock: disco key = d:e643e2b83d025f05
2024/12/10 23:40:55 Creating WireGuard device...
2024/12/10 23:40:55 Bringing WireGuard device up...
2024/12/10 23:40:55 external route: up
2024/12/10 23:40:55 Bringing router up...
2024/12/10 23:40:55 Clearing router settings...
2024/12/10 23:40:55 Starting network monitor...
2024/12/10 23:40:55 Engine created.
2024/12/10 23:40:55 monitor: [unexpected] network state changed, but stringification didn't: interfaces.State{defaultRoute=eth0 ifs={eth0:[172.19.0.2/16]} v4=true v6=false}
2024/12/10 23:40:55 monitor: [unexpected] old: {"InterfaceIPs":{"eth0":["172.19.0.2/16"],"lo":["127.0.0.1/8","::1/128"]},"Interface":{"eth0":{"Index":439,"MTU":1500,"Name":"eth0","HardwareAddr":"AkKsEwAC","Flags":51,"AltAddrs":null,"Desc":""},"lo":{"Index":1,"MTU":65536,"Name":"lo","HardwareAddr":null,"Flags":37,"AltAddrs":null,"Desc":""}},"HaveV6":false,"HaveV4":true,"Is
Expensive":false,"DefaultRouteInterface":"eth0","HTTPProxy":"","PAC":""}
2024/12/10 23:40:55 monitor: [unexpected] new: {"InterfaceIPs":{"eth0":["172.19.0.2/16"],"lo":["127.0.0.1/8","::1/128"],"tailscale0":["fe80::eb7:4af9:916d:8442/64"]},"Interface":{"eth0":{"Index":439,"MTU":1500,"Name":"eth0","HardwareAddr":"AkKsEwAC","Flags":51,"AltAddrs":null,"Desc":""},"lo":{"Index":1,"MTU":65536,"Name":"lo","HardwareAddr":null,"Flags":37,"AltAddrs":null,"Desc":""},"tailscale0":{"Index":2,"MTU":1280,"Name":"tailscale0","HardwareAddr":null,"Flags":57,"AltAddrs":null,"Desc":""}},"HaveV6":false,"HaveV4":true,"IsExpensive":false,"DefaultRouteInterface":"eth0","HTTPProxy":"","PAC":""}
2024/12/10 23:40:55 LinkChange: major, rebinding. New state: interfaces.State{defaultRoute=eth0 ifs={eth0:[172.19.0.2/16]} v4=true v6=false}
2024/12/10 23:40:55 onPortUpdate(port=41366, network=udp6)
2024/12/10 23:40:55 onPortUpdate(port=42519, network=udp4)
2024/12/10 23:40:55 Rebind; defIf="eth0", ips=[172.19.0.2/16]
2024/12/10 23:40:55 magicsock: 0 active derp conns
2024/12/10 23:40:55 pm: using backend prefs for "profile-5f86": Prefs{ra=false dns=false want=true routes=[] statefulFiltering=false tags=tag:container nf=on host="uptime" update=check Persist{lm=, o=, n=[Em+2Z] u="uptime.astrapia-mermaid.ts.net"}}
2024/12/10 23:40:55 logpolicy: using system state directory "/var/lib/tailscale"
2024/12/10 23:40:55 monitor: gateway and self IP changed: gw=172.19.0.1 self=172.19.0.2
2024/12/10 23:40:55 got LocalBackend in 28ms
2024/12/10 23:40:55 Start
2024/12/10 23:40:55 Backend: logs: be:b81787e5ddb79b647317d01a7e310fad69865d61b22803ccdba8aea3e19d4cbc fe:
2024/12/10 23:40:55 control: client.Login(0)
2024/12/10 23:40:55 health(warnable=warming-up): error: Tailscale is starting. Please wait.
2024/12/10 23:40:55 control: doLogin(regen=false, hasUrl=false)
boot: 2024/12/10 23:40:55 [warning] failed to symlink socket: file exists
        To interact with the Tailscale CLI please use `tailscale --socket="/tmp/tailscaled.sock"`
boot: 2024/12/10 23:40:55 Running 'tailscale up'
2024/12/10 23:40:55 control: control server key from https://controlplane.tailscale.com: ts2021=[fSeS+], legacy=[nlFWp]
2024/12/10 23:40:55 control: RegisterReq: onode= node=[Em+2Z] fup=false nks=false
2024/12/10 23:40:55 control: RegisterReq: got response; nodeKeyExpired=false, machineAuthorized=true; authURL=false
2024/12/10 23:40:55 control: netmap: got new dial plan from control
2024/12/10 23:40:55 health(warnable=not-in-map-poll): ok
2024/12/10 23:40:55 active login: uptime.astrapia-mermaid.ts.net
2024/12/10 23:40:55 Switching ipn state NoState -> Starting (WantRunning=true, nm=true)
2024/12/10 23:40:55 magicsock: SetPrivateKey called (init)
2024/12/10 23:40:55 wgengine: Reconfig: configuring userspace WireGuard config (with 0/12 peers)
2024/12/10 23:40:55 wgengine: Reconfig: configuring router
2024/12/10 23:40:55 Start
2024/12/10 23:40:55 active login: 
2024/12/10 23:40:55 Backend: logs: be:b81787e5ddb79b647317d01a7e310fad69865d61b22803ccdba8aea3e19d4cbc fe:
2024/12/10 23:40:55 control: client.Login(0)
2024/12/10 23:40:55 control: client.Shutdown ...
2024/12/10 23:40:55 control: doLogin(regen=false, hasUrl=false)
2024/12/10 23:40:55 control: updateRoutine: exiting
2024/12/10 23:40:55 control: authRoutine: exiting
2024/12/10 23:40:55 control: mapRoutine: exiting
2024/12/10 23:40:55 wgengine: Reconfig: configuring DNS
2024/12/10 23:40:55 dns: Set: {DefaultResolvers:[] Routes:{} SearchDomains:[] Hosts:14}
2024/12/10 23:40:55 dns: Resolvercfg: {Routes:{} Hosts:14 LocalDomains:[]}
2024/12/10 23:40:55 dns: OScfg: {}
2024/12/10 23:40:55 control: Client.Shutdown done.
2024/12/10 23:40:55 control: control server key from https://controlplane.tailscale.com: ts2021=[fSeS+], legacy=[nlFWp]
2024/12/10 23:40:55 control: RegisterReq: onode= node=[Em+2Z] fup=false nks=false
2024/12/10 23:40:55 control: controlhttp: forcing port 443 dial due to recent noise dial
2024/12/10 23:40:55 magicsock: home DERP changing from derp-0 [0ms] to derp-14 [35ms]
2024/12/10 23:40:55 magicsock: home is now derp-14 (ams)
2024/12/10 23:40:55 magicsock: adding connection to derp-14 for home-keep-alive
2024/12/10 23:40:55 magicsock: 1 active derp conns: derp-14=cr0s,wr0s
2024/12/10 23:40:55 derphttp.Client.Connect: connecting to derp-14 (ams)
2024/12/10 23:40:55 magicsock: endpoints changed: 195.52.157.27:42519 (stun), 172.19.0.2:42519 (local)
2024/12/10 23:40:55 health(warnable=no-derp-connection): ok
2024/12/10 23:40:55 control: NetInfo: NetInfo{varies=false hairpin= ipv6=false ipv6os=true udp=true icmpv4=false derp=#14 portmap= link="" firewallmode="ipt-default"}
2024/12/10 23:40:55 health(warnable=no-derp-connection): ok
2024/12/10 23:40:55 magicsock: derp-14 connected; connGen=1
2024/12/10 23:40:55 health(warnable=no-derp-connection): ok
2024/12/10 23:40:55 health(warnable=no-derp-connection): ok
2024/12/10 23:40:55 [RATELIMIT] format("health(warnable=%s): ok")
2024/12/10 23:40:55 control: controlhttp: forcing port 443 dial due to recent noise dial
2024/12/10 23:40:56 control: controlhttp: forcing port 443 dial due to recent noise dial
2024/12/10 23:40:56 control: RegisterReq: got response; nodeKeyExpired=false, machineAuthorized=true; authURL=false
2024/12/10 23:40:56 control: netmap: got new dial plan from control
2024/12/10 23:40:56 active login: uptime.astrapia-mermaid.ts.net
2024/12/10 23:40:56 Switching ipn state NoState -> Starting (WantRunning=true, nm=true)
2024/12/10 23:40:56 peerapi: serving on http://100.95.31.16:44645
2024/12/10 23:40:56 peerapi: serving on http://[fd7a:115c:a1e0::e501:1f10]:44511
2024/12/10 23:40:56 Switching ipn state Starting -> Running (WantRunning=true, nm=true)
boot: 2024/12/10 23:40:56 serve proxy: unsetting previous config
2024/12/10 23:40:56 tailscaled got signal terminated; shutting down
2024/12/10 23:40:56 canceling captive portal context
2024/12/10 23:40:56 control: client.Shutdown ...
2024/12/10 23:40:56 control: updateRoutine: exiting
2024/12/10 23:40:56 control: authRoutine: exiting
2024/12/10 23:40:56 control: mapRoutine: exiting
2024/12/10 23:40:56 control: Client.Shutdown done.
2024/12/10 23:40:56 magicsock: closing connection to derp-14 (conn-close), age 0s
2024/12/10 23:40:56 magicsock: 0 active derp conns
2024/12/10 23:40:56 monitor: RTM_DELROUTE: src=, dst=fe80::/64, gw=, outif=2, table=254
2024/12/10 23:40:56 monitor: RTM_DELROUTE: src=, dst=fe80::eb7:4af9:916d:8442/128, gw=, outif=2, table=255
2024/12/10 23:40:56 monitor: RTM_DELROUTE: src=, dst=ff00::/8, gw=, outif=2, table=255
2024/12/10 23:40:56 monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:254 Protocol:0 Scope:0 Type:1 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5210 Table:254 Mark:16711680 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
2024/12/10 23:40:56 monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:253 Protocol:0 Scope:0 Type:1 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5230 Table:253 Mark:16711680 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
2024/12/10 23:40:56 monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:0 Protocol:0 Scope:0 Type:7 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5250 Table:0 Mark:16711680 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
2024/12/10 23:40:56 monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:52 Protocol:0 Scope:0 Type:1 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5270 Table:52 Mark:0 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa72e81]

goroutine 1 [running]:
main.(*kubeClient).storeHTTPSEndpoint(...)
        tailscale.com/cmd/containerboot/kube.go:88
main.main()
        tailscale.com/cmd/containerboot/main.go:334 +0x14e1
boot: 2024/12/10 23:41:06 Starting tailscaled
boot: 2024/12/10 23:41:06 Waiting for tailscaled socket
2024/12/10 23:41:06 logtail started
2024/12/10 23:41:06 Program starting: v1.76.6-t1edcf9d46, Go 1.23.1: []string{"tailscaled", "--socket=/tmp/tailscaled.sock", "--statedir=/var/lib/tailscale"}
```

### Steps to reproduce

- Create a docker-compose with a Tailscale sidecar, example Uptime-Kuma
- Configure an OAuth client as described in the Wiki
- Use  `image: tailscale/tailscale:latest` as a definition and try to start the service, it should result in log messages similar to the above
- Change to use `image: tailscale/tailscale:v1.76` and restart, the service can be used successfully

### Are there any recent changes that introduced the issue?

- https://github.com/tailscale/tailscale/releases/tag/v1.78.1 released 5 days ago, no problems
- https://hub.docker.com/layers/tailscale/tailscale/latest/images/sha256-69203be03a0d30bd23acaccdecfbfb41cd21d9e4997e305ad47b881e4e0acaa3?context=explore released 5 hours ago, multiple services went down at the same time

### OS

Linux

### OS version

Ubuntu 24.04

### Tailscale version

1.78.1

### Other software

```
$ docker --version
Docker version 27.2.0, build 3ab4256
$ snap list docker 
Name    Version  Rev   Tracking       Publisher   Notes
docker  27.2.0   2963  latest/stable  canonical✓  -
```

### Bug report

BUG-19f00151c3dcba4060a76fae5ed515aae9f201e6fdcf815d82509711b05a972b-20241211002800Z-39d71d4a1925f2fe

--- a/cmd/containerboot/main.go
+++ b/cmd/containerboot/main.go
@@ -331,8 +331,10 @@ authLoop:
 		if err := client.SetServeConfig(ctx, new(ipn.ServeConfig)); err != nil {
 			log.Fatalf("failed to unset serve config: %v", err)
 		}
-		if err := kc.storeHTTPSEndpoint(ctx, ""); err != nil {
-			log.Fatalf("failed to update HTTPS endpoint in tailscale state: %v", err)
+		if hasKubeStateStore(cfg) {
+			if err := kc.storeHTTPSEndpoint(ctx, ""); err != nil {
+				log.Fatalf("failed to update HTTPS endpoint in tailscale state: %v", err)
+			}
 		}
 	}
 

